luzer: fix memory leak in FDP

ligurio opened the pull request:
Fixes #52
Fixes #65
Fixes #85

ligurio committed:
luzer: fix memory leaks

The patch fixes memory leaks in luzer that manifest themselves in
different ways:

1. `LLVMFuzzerRunDriver()` returns due to error with dictionary.
2. Lua objects accumulate memory between `TestOneInput()` runs,

but the cause is the same for everyone: Lua is a language with
GC-based memory management, and after finishing the work, we
did not free the memory occupied by Lua objects, so libFuzzer
thinks there was a memory leak.

1. https://llvm.org/docs/LibFuzzer.html#options

Fixes #52
Fixes #65

ligurio committed:
luzer: fix memory leaks

The patch fixes memory leaks in luzer that manifest themselves in
different ways:

1. `LLVMFuzzerRunDriver()` returns due to error with dictionary.
2. Lua objects accumulate memory between `TestOneInput()` runs,

but the cause is the same for everyone: Lua is a language with
GC-based memory management, and after finishing the work, we
did not free the memory occupied by Lua objects, so libFuzzer
thinks there was a memory leak.

1. https://llvm.org/docs/LibFuzzer.html#options

Fixes #52
Fixes #65
Fixes #85

Buristan reviewed (changes requested):
Sergey,
Thanks for the patch!
I suppose the next commit may be squashed into this one as well.

Please consider my comment below.

ligurio commented on luzer/luzer.c:
```diff
@@ -345,6 +345,9 @@ TestOneInput(const uint8_t* data, size_t size) {
 	LUA_SETHOOK(L, debug_hook, 0, 0);
 
+	/* Prevents memory leaks on module exit. */
+	lua_gc(L, LUA_GCCOLLECT, 0);
```
Also https://github.com/google/atheris/blob/master/native_extension_fuzzing.md#leak-detection

> Leak detection
> Python is known to leak certain data, such as at interpreter initialization time. You should disable leak detection, for example with -detect_leaks=0.

ligurio committed:
xxx

ligurio committed:
cmake: FORTIFY_SOURCE

Проблема: На NixOS clang wrapper через hardening flags добавляет -D_FORTIFY_SOURCE=2 после пользовательских флагов, а -O2 — до. Когда CMake переопределяет -O2 на -O0 (для ASan/UBSan тестовых целей), -D_FORTIFY_SOURCE=2 остаётся, и glibc 2.40+ выдаёт ошибку: _FORTIFY_SOURCE requires compiling with optimization (-O).

Исправление (в двух файлах):

1. luzer/CMakeLists.txt:46-52 — Добавлен -U_FORTIFY_SOURCE в общие флаги компиляции (belt-and-suspenders).
2. luzer/tests/CMakeLists.txt:218-219, 350-351 — Заменены -O0 на -O1 для ASan/UBSan тестовых целей (luac_asan, luac_ubsan, testlib_asan, testlib_ubsan). -O1 — минимальный рекомендуемый уровень оптимизации для санитайзеров.

ligurio committed:
leak fdp repro

ligurio committed:
Revert "cmake: FORTIFY_SOURCE"

This reverts commit 41667f6731bd18ffb4475156d153d40c93766b07.
